xarray/test/test_dataarray.py

```diff
@@ -63,6 +66,67 @@ def test_data_property(self):
         self.assertArrayEqual(2 * np.ones((3, 4)), actual.data)
         self.assertArrayEqual(actual.data, actual.values)
 
+    def test_struct_array_dims(self):
+        # -----------------test 1
+        p_data = np.array([('John', 180), ('Stacy', 150), ('Dick', 200)],
+                          dtype=[('name', '|S256'), ('height', object)])
+        p_data_1 = np.array([('John', 180), ('Stacy', 150), ('Dick', 200)],
+                            dtype=[('name', '|S256'), ('height', object)])
+
+        weights_0 = DataArray([80, 56, 120], dims=['participant'],
+                              coords={'participant': p_data})
+        weights_1 = DataArray([81, 52, 115], dims=['participant'],
+                              coords={'participant': p_data_1})
+
+        actual = weights_1 - weights_0
+
+        expected = DataArray([1, -4, -5], dims=['participant'],
+                             coords={'participant': p_data})
+
+        self.assertDataArrayIdentical(actual, expected)
+
+        # -----------------test 2
+        p_data = np.array([('John', 180), ('Stacy', 150), ('Dick', 200)],
+                          dtype=[('name', '|S256'), ('height', object)])
+        p_data_1 = np.array([('John', 180), ('Stacy', 151), ('Dick', 200)],
+                            dtype=[('name', '|S256'), ('height', object)])
+
+        p_data_2 = np.array([('John', 180), ('Dick', 200)],
+                            dtype=[('name', '|S256'), ('height', object)])
+
+        weights_0 = DataArray([80, 56, 120], dims=['participant'],
+                              coords={'participant': p_data})
+        weights_1 = DataArray([81, 52, 115], dims=['participant'],
+                              coords={'participant': p_data_1})
+
+        actual = weights_1 - weights_0
+
+        expected = DataArray([1, -5], dims=['participant'],
+                             coords={'participant': p_data_2})
+
+        self.assertDataArrayIdentical(actual, expected)
+
+        # -----------------test 3 - np.nan
+        p_data = np.array([('John', 180), ('Stacy', 150), ('Dick', 200)],
+                          dtype=[('name', '|S256'), ('height', object)])
+        p_data_1 = np.array([('John', 180), ('Stacy', np.nan), ('Dick', 200)],
+                            dtype=[('name', '|S256'), ('height', object)])
+
+        p_data_2 = np.array([('John', 180), ('Dick', 200)],
+                            dtype=[('name', '|S256'), ('height', object)])
+
+        weights_0 = DataArray([80, 56, 120], dims=['participant'],
+                              coords={'participant': p_data})
+        weights_1 = DataArray([81, 52, 115], dims=['participant'],
+                              coords={'participant': p_data_1})
+
+        actual = weights_1 - weights_0
+
+        expected = DataArray([1, -5], dims=['participant'],
+                             coords={'participant': p_data_2})
+
+        self.assertDataArrayIdentical(actual, expected)
+
```

Comment:
I think the tests you've written are overly verbose.  I don't think you need to recreate `p_data*` and `weights*` for each of your tests.  The way it is written now, I'm not sure what the differences between your three tests are (at least test1 and test2).